xfrmi: fix xfrm interface ip refcount problem

wofferl opened the pull request:
As mentioned in https://github.com/libreswan/libreswan/issues/1778#issuecomment-2285548428 there is actual an issue with referencing xfrm interfacip.

Actual add_xfrm_interface is called in several places when the connection is going up and add_xfrm_interface_ip which create/reference the interface ip is also called from there.
But unreference_xfrmi_ip is only called from unreference_xfrmi when the connection is deleted (discard_connection).

Previous in e727e23 double ip references were removed and in 6b66fc5 a new double reference added.

This leads to the following problems:

* if a connection is deleted that had never been up, it regardless unreferencing an ip. Which leads either to an error "Unable to unreference IP on xfrmi device" or it deletes the ip from another connection as described in #1778
* for every up/down an new reference is created, but the ip will never been deleted

So moving the referencing/unreferencing to updown depending on UPDOWN_ROUTE/UPDOWN_UNROUTE makes sure the ip is correctly referenced for routed-ondemand and routed-tunnel.

I added also an fix for the initial xfrm interface check.

cagney commented:
The principal seems sound.

I'm just not sure that do_updown() is the right place.  Having a function running scripts also add/delete interfaces is a little quirky.

wofferl commented:
> The principal seems sound.
> 
> I'm just not sure that do_updown() is the right place. Having a function running scripts also add/delete interfaces is a little quirky.

An alternative could be moving them to routing.c/kernel.c before the specific do_updown calls, with the risk to overlook it making changes in the future:

grep -ri "do_updown.*route" programs/

```
programs/pluto/kernel.c:                                do_updown(UPDOWN_ROUTE, c, spd, NULL/*state*/, c->logger);
programs/pluto/kernel.c:                                do_updown(UPDOWN_ROUTE, c, spd, child, logger);
programs/pluto/routing.c:       if (!do_updown(UPDOWN_ROUTE, c, c->spd, NULL/*ST*/, logger)) {
programs/pluto/routing.c:               do_updown_unroute_spd(spd, &owner, NULL, logger);
programs/pluto/routing.c:               do_updown_unroute_spd(spd, &owner, child, logger);
programs/pluto/routing.c:       do_updown_child(UPDOWN_ROUTE, child);
programs/pluto/routing.c:       do_updown_child(UPDOWN_ROUTE, child);
programs/pluto/routing.c:               do_updown_unroute_spd(spd, &owner, child, logger);
programs/pluto/routing.c:               do_updown_unroute(c, (*e->child));
programs/pluto/routing.c:               do_updown_child(UPDOWN_ROUTE, (*e->child));
programs/pluto/updown.c:void do_updown_unroute_spd(const struct spd *spd, const struct spd_owner *owner,
programs/pluto/updown.c:                do_updown(UPDOWN_UNROUTE, spd->connection,
programs/pluto/updown.c:void do_updown_unroute(const struct connection *c, struct child_sa *child)
programs/pluto/updown.c:                do_updown_unroute_spd(spd, &owner, child, logger);
programs/pluto/updown.h:void do_updown_unroute(const struct connection *c, struct child_sa *child);
programs/pluto/updown.h:void do_updown_unroute_spd(const struct spd *spd, const struct spd_owner *owner,
```

cagney commented:
> An alternative could be moving them to routing.c/kernel.c before the specific do_updown calls, with the risk to overlook it making changes in the future

Right.

Digging some more, I noticed:

- MOBIKE, for SUSPEND, calls do_updown(UNROUTE) on the ROUTED_TUNNEL, followed later by RESUME calling do_updown(ROUTE)
Would this release the assigned ipsec-interface?
- do_updown() is run per-SPD (this is arguably a bug), but the ipsec-interface is per-connection
The effect is to give the connection a reference per SPD.  This might make it it easier (for instance, when things fail part way through the SPDs), I'm just not sure how well things go when there's narrowing and the SPDs get re-built.

Two other thoughts:

- add something like do_{route,unroute}() which both adjusts the refcount and calls do_updown(*same)
i.e., invert the problem
- hook dispatch() to, when the transition is to unrouted, do the call
I suspect this is something for the next iteration of the state machine

cagney commented:
Would things be conceptually simpler if UNROUTED->* assigned the ipsec-interface, and *->UNROUTED released it?

wofferl commented:
If it's consistent, yes. But I see actual some problems about this. 

I added a log for me to understand the states better:
```
diff --git a/programs/pluto/routing.c b/programs/pluto/routing.c
index f21c1ec130..498bd49bf7 100644
--- a/programs/pluto/routing.c
+++ b/programs/pluto/routing.c
@@ -1642,6 +1642,7 @@ static bool dispatch_1(enum routing_event event,
        const enum routing routing = c->routing.state;
        const enum connection_kind kind = c->local->kind;

+       llog(RC_ROUTE, logger, "DEBUG show events : %s->%s",routing_event_name[routing], routing_event_name[event]);
        switch (XX(event, routing, kind)) {

        case X(ROUTE, UNROUTED, GROUP):
```

The connection state is initially in state CONNECTION_ROUTE without going through dispatch_1 (routing.c when it is added and it would unrouted twice when it is unrouted and deleted. Maybe I'm thinking wrong here or are in the wrong place.

[build@libreswan-testsuite OUTPUT]$ sudo ipsec add eastnet-northnet
`"eastnet-northnet": added IKEv2 connection`
[build@libreswan-testsuite OUTPUT]$ sudo ipsec unroute eastnet-northnet
`"eastnet-northnet": DEBUG show events : CONNECTION_ROUTE->CONNECTION_UNROUTE`
[build@libreswan-testsuite OUTPUT]$ sudo ipsec delete eastnet-northnet
`"eastnet-northnet": DEBUG show events : CONNECTION_ROUTE->CONNECTION_UNROUTE`

Also multiple ipsec add, route or unroute have unwanted behaviour.

[build@libreswan-testsuite OUTPUT]$ sudo ipsec add eastnet-northnet
`"eastnet-northnet": added IKEv2 connection`
[build@libreswan-testsuite OUTPUT]$ sudo ipsec add eastnet-northnet
`"eastnet-northnet": DEBUG show events : CONNECTION_ROUTE->CONNECTION_UNROUTE`
`"eastnet-northnet": added IKEv2 connection`

[build@libreswan-testsuite OUTPUT]$ sudo ipsec unroute eastnet-northnet
`"eastnet-northnet": DEBUG show events : CONNECTION_ROUTE->CONNECTION_UNROUTE`
[build@libreswan-testsuite OUTPUT]$ sudo ipsec unroute eastnet-northnet
`"eastnet-northnet": DEBUG show events : CONNECTION_ROUTE->CONNECTION_UNROUTE`

wofferl commented:
If it is consistent then yes, as long as there is no duplication for example "start UNROUTE, UNROUTED, PERMANENT;", when you call unroute multiple times, or unroute/delete.

Don't know if there are any other states that needed to be excluded.

cagney commented:
btw, when  moving the code to `updown.c` the check:
```
	if (c->xfrmi != NULL && c->xfrmi->if_id != 0)
```
seems to have been lost

cagney commented:
See XXX comment:
```
diff --git a/programs/pluto/routing.c b/programs/pluto/routing.c
index c777c84d44..7cb53733dc 100644
--- a/programs/pluto/routing.c
+++ b/programs/pluto/routing.c
@@ -31,6 +31,9 @@
 #include "updown.h"
 #include "instantiate.h"
 #include "connection_event.h"
+#ifdef USE_XFRM_INTERFACE
+#include "kernel_xfrm_interface.h"
+#endif
 
 enum routing_event {
 	/* fiddle with the ROUTE bit */
@@ -2444,6 +2447,44 @@ static bool dispatch(enum routing_event event,
 			if (e->dispatch_ok == NULL ||
 			    e->dispatch_ok(c, logger, e)) {
 				ok = dispatch_1(event, c, logger, e);
+#ifdef USE_XFRM_INTERFACE
+				/*
+				 * Whenever a connection transitions
+				 * out of or into UNROUTED, add/remove
+				 * the ipsec_interface..
+				 *
+				 * XXX: is this too late?
+				 *
+				 * No!  It's too early.
+				 *
+				 * ikev1-xfrmi-05-remote-access-client,
+				 * for instance, fails with:
+				 *
+				 * doroute "ip -4 route replace
+				 * 192.0.2.0/24 dev ipsec1 src
+				 * 192.0.2.209" failed (Error: Invalid
+				 * prefsrc address.)
+				 *
+				 * Because the address wasn't set on
+				 * the interface when it was created
+				 * (the peer hasn't yet sent the
+				 * address).
+				 */
+				if (old.routing != c->routing.state &&
+				    c->xfrmi != NULL && c->xfrmi->if_id != 0) {
+					PEXPECT(logger, ok);
+					if (c->routing.state == RT_UNROUTED) {
+						ldbg_routing(logger, "removing ipsec-interface %"PRIu32,
+							     c->xfrmi->if_id);
+						remove_xfrm_interface(c, logger);
+					}
+					if (old.routing == RT_UNROUTED) {
+						ldbg_routing(logger, "adding ipsec-interface %"PRIu32,
+							     c->xfrmi->if_id);
+						ok = add_xfrm_interface(c, logger);
+					}
+				}
+#endif
 			}
 			if (ok && e->post_op != NULL) {
 				e->post_op(e);

```

cagney commented:
@wofferl your changes result in this:
```
[cagney@bernard git-endpoint main↑2]$ ./kvm diff testing/pluto/ikev2-xfrmi-12-two-default-routes
--- MASTER/testing/pluto/ikev2-xfrmi-12-two-default-routes/east.console.txt
+++ OUTPUT/testing/pluto/ikev2-xfrmi-12-two-default-routes/east.console.txt
@@ -13,6 +13,6 @@
 initdone
 east #
  ipsec whack --trafficstatus
-#2: "north-east-default", type=ESP, add_time=1234567890, inBytes=84, outBytes=84, maxBytes=2^63B, id='@north'
+#2: "north-east-default", type=ESP, add_time=1234567890, inBytes=0, outBytes=0, maxBytes=2^63B, id='@north'
 east #
  
--- MASTER/testing/pluto/ikev2-xfrmi-12-two-default-routes/north.console.txt
+++ OUTPUT/testing/pluto/ikev2-xfrmi-12-two-default-routes/north.console.txt
@@ -71,26 +71,25 @@
            336       4      0       0       0       0 
 north #
  ../../guestbin/ip.sh route add 192.0.2.0/24 dev ipsec3
+Cannot find device "ipsec3"
 north #
  ../../guestbin/ping-once.sh --up 192.0.2.254
-up
+down UNEXPECTED
+# ping -n -c 1  -i 6 -w 5    192.0.2.254
+PING 192.0.2.254 (192.0.2.254) 56(84) bytes of data. --- 192.0.2.254 ping statistics --- 1 packets transmitted, 0 received, 100% packet loss, time XXXX
 north #
  ../../guestbin/ip.sh -s link show ipsec3
-X: ipsec3@eth1: <NOARP,UP,LOWER_UP> mtu 1500 state UNKNOWN
-    RX:  bytes packets errors dropped  missed   mcast           
-            84       1      0       0       0       0 
-    TX:  bytes packets errors dropped carrier collsns           
-            84       1      0       0       0       0 
+Device "ipsec3" does not exist.
 north #
  ipsec trafficstatus
-#4: "north-east", type=ESP, add_time=1234567890, inBytes=84, outBytes=84, maxBytes=2^63B, id='@east'
+#4: "north-east", type=ESP, add_time=1234567890, inBytes=0, outBytes=0, maxBytes=2^63B, id='@east'
 #2: "north-west", type=ESP, add_time=1234567890, inBytes=336, outBytes=336, maxBytes=2^63B, id='@west'
 north #
  echo "initdone"
 initdone
 north #
  ipsec whack --trafficstatus
-#4: "north-east", type=ESP, add_time=1234567890, inBytes=84, outBytes=84, maxBytes=2^63B, id='@east'
+#4: "north-east", type=ESP, add_time=1234567890, inBytes=0, outBytes=0, maxBytes=2^63B, id='@east'
 #2: "north-west", type=ESP, add_time=1234567890, inBytes=336, outBytes=336, maxBytes=2^63B, id='@west'
 north #
  
```

cagney commented:
and
```
[cagney@bernard git-endpoint main↑2]$ ./kvm diff testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi
--- MASTER/testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi/west.console.txt
+++ OUTPUT/testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi/west.console.txt
@@ -62,15 +62,11 @@
  ip --color=never link show ipsec1
 Device "ipsec1" does not exist.
 west #
- 
-west #
  #
 west #
  # invalid device; <<ipsec add>> is before <<ip link>>
 west #
  #
-west #
- 
 west #
  ipsec add westnet4-eastnet4
 "westnet4-eastnet4": added IKEv2 connection
@@ -123,8 +119,8 @@
  ipsec delete westnet6-eastnet6
 "westnet6-eastnet6": terminating SAs using this connection
 "westnet6-eastnet6" #8: deleting IKE SA (ESTABLISHED_IKE_SA) and sending notification
+ERROR: "westnet6-eastnet6" #10: kernel: xfrm XFRM_MSG_DELPOLICY delete response for flow (in): No such file or directory (errno 2)
 ERROR: "westnet6-eastnet6" #10: unreference_xfrmi_ip() Unable to unreference IP on xfrmi device [ipsec1] id [1]
-"westnet6-eastnet6" #10: ESP traffic information: in=0B out=0B
 west #
  ip --color=never link show ipsec1
 X: ipsec1@NONE: <NOARP> mtu 1500 qdisc state DOWN qlen 1000
```

wofferl commented:
So actual I'm thinking of this, call add_xfrm_interface when:
* RT_UNROUTED->RT_ROUTE_ONDEMAND
*  RT_UNROUTED_INBOUND->RT_TUNNEL

Not sure if an additional query **event=CONNECTION_ESTABLISH_INBOUND/OUTBOUND** would be needed for the second one.

```
diff --git a/programs/pluto/routing.c b/programs/pluto/routing.c
index c777c84d44..55aa725fc7 100644
--- a/programs/pluto/routing.c
+++ b/programs/pluto/routing.c
@@ -31,6 +31,9 @@
 #include "updown.h"
 #include "instantiate.h"
 #include "connection_event.h"
+#ifdef USE_XFRM_INTERFACE
+#include "kernel_xfrm_interface.h"
+#endif

 enum routing_event {
        /* fiddle with the ROUTE bit */
@@ -2441,9 +2444,35 @@ static bool dispatch(enum routing_event event,
        {
                struct old_routing old = ldbg_routing_start(event, c, logger, e);
                {
+#ifdef USE_XFRM_INTERFACE
+                       if(c->xfrmi != NULL && c->xfrmi->if_id != 0) {
+                               PEXPECT(logger, ok);
+                               /*
+                                *  add xfrm interface early for route on demand RT_UNROUTED->RT_ROUTE_ONDEMAND
+                                *  or when established from unrouted tunnel RT_UNROUTED_INBOUND->RT_TUNNEL
+                                */
+                               if ((c->routing.state == RT_UNROUTED && event == CONNECTION_ROUTE) ||
+                                    c->routing.state == RT_UNROUTED_INBOUND) {
+                                       ldbg_routing(logger, "adding ipsec-interface %"PRIu32,
+                                                    c->xfrmi->if_id);
+                                       ok = add_xfrm_interface(c, logger);
+                               }
+                       }
+#endif
                        if (e->dispatch_ok == NULL ||
                            e->dispatch_ok(c, logger, e)) {
                                ok = dispatch_1(event, c, logger, e);
+#ifdef USE_XFRM_INTERFACE
+                               if (old.routing != c->routing.state &&
+                                   c->xfrmi != NULL && c->xfrmi->if_id != 0) {
+                                       PEXPECT(logger, ok);
+                                       if (c->routing.state == RT_UNROUTED) {
+                                               ldbg_routing(logger, "removing ipsec-interface %"PRIu32,
+                                                            c->xfrmi->if_id);
+                                               remove_xfrm_interface(c, logger);
+                                       }
+                               }
+#endif
                        }
                        if (ok && e->post_op != NULL) {
                                e->post_op(e);
```

cagney commented:
Heads up, this:
```
+#ifdef USE_XFRM_INTERFACE
+#include "kernel_xfrm_interface.h"
+#endif
```
is being replaced by:
```
#include "ipsec_interface.h"
```
C code calling functions still needs #ifdef USE_XFRM_INTERFACE.

wofferl commented:
I commited my last change as described in https://github.com/libreswan/libreswan/pull/1795#issuecomment-2346144859

Not entirely happy with the query for add_xfrm_interface, but all tests looks good and it handles the interface correctly.

Also not sure what to do when add_xfrm_interface fails. 

`ok = add_xfrm_interface(c, logger);`

as you mentioned in https://github.com/libreswan/libreswan/pull/1795#issuecomment-2344433249 seems not optimal, as the test  ipsec-interface-15-blocked-ikev2-xfrmi] shows

```
 ipsec up westnet4-eastnet4
"westnet4-eastnet4" #5: initiating IKEv2 connection to 192.1.2.23 using UDP
"westnet4-eastnet4" #5: sent IKE_SA_INIT request to 192.1.2.23:UDP/500
"westnet4-eastnet4" #5: processed IKE_SA_INIT response from 192.1.2.23:UDP/500 {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=DH19}, initiating IKE_AUTH
"westnet4-eastnet4" #5: sent IKE_AUTH request to 192.1.2.23:UDP/500
"westnet4-eastnet4" #5: initiator established IKE SA; authenticated peer using authby=secret and ID_FQDN '@east'
ERROR: "westnet4-eastnet4" #X: ipsec1 has an xfrm device ifindex (RTA_LINK) of 0
ERROR: "westnet4-eastnet4" #6: device ipsec1 exists but do not match expected type, XFRM if_id 1, or XFRM device is invalid; check 'ip -d link show dev ipsec1'
"westnet4-eastnet4" #6: route-client output: PATH/libexec/ipsec/_updown.xfrm: doroute "ip -4 route replace 192.0.2.0/24   dev ipsec1" failed (Error: Device for nexthop is not up.)
"westnet4-eastnet4" #6: initiator established Child SA using #5; IPsec tunnel [192.0.1.0/24===192.0.2.0/24] {ESP/ESN=>0xESPESP <0xESPESP xfrm=AES_GCM_16_256-NONE DPD=passive}
```

```
west #
 ipsec up westnet6-eastnet6
"westnet6-eastnet6" #7: initiating IKEv2 connection to 2001:db8:1:2::23 using UDP
"westnet6-eastnet6" #7: sent IKE_SA_INIT request to [2001:db8:1:2::23]:UDP/500
"westnet6-eastnet6" #7: processed IKE_SA_INIT response from [2001:db8:1:2::23]:UDP/500 {cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=DH19}, initiating IKE_AUTH
"westnet6-eastnet6" #7: sent IKE_AUTH request to [2001:db8:1:2::23]:UDP/500
"westnet6-eastnet6" #7: initiator established IKE SA; authenticated peer using authby=secret and ID_FQDN '@east'
ERROR: "westnet6-eastnet6" #X: ipsec1 has an xfrm device ifindex (RTA_LINK) of 0
ERROR: "westnet6-eastnet6" #8: device ipsec1 exists but do not match expected type, XFRM if_id 1, or XFRM device is invalid; check 'ip -d link show dev ipsec1'
"westnet6-eastnet6" #8: route-client-v6 output: PATH/libexec/ipsec/_updown.xfrm: doroute "ip -6 route replace 2001:db8:0:2::/64   dev ipsec1" failed (Error: Nexthop device is not up.)
"westnet6-eastnet6" #8: route-client-v6 command exited with status 2
EXPECTATION FAILED: "westnet6-eastnet6" #8: ok (dispatch() +2466 programs/pluto/routing.c)
"westnet6-eastnet6" #8: CHILD SA failed: TEMPORARY_FAILURE
"westnet6-eastnet6" #7: IKE SA established but initiator rejected Child SA response
"westnet6-eastnet6" #8: sent INFORMATIONAL request to delete larval Child SA using IKE SA #7
"westnet6-eastnet6" #8: connection is supposed to remain up; revival attempt 1 scheduled in 0 seconds
ERROR: "westnet6-eastnet6" #8: kernel: xfrm XFRM_MSG_DELPOLICY delete response for flow (in): No such file or directory (errno 2)
"westnet6-eastnet6" #8: kernel: replace_ipsec_with_bare_kernel_policy() inbound delete failed
```

cagney commented:
I'm testing; even with the quirk you noted it is way better.
